### PR TITLE
Unprocessed batch items with recursive backoff

### PR DESCRIPTION
**Pull request by nhyne**
I haven't reviewed this with John yet, I was hoping to do that next week assuming he has time. 

I was initially hoping to have the `BatchWriteItem` case class take a schedule instead of just a number of retries and the time to wait. I had some problems with using the schedule however, primarily that the schedule helper on `ZIO` returns the schedule's value rather than the ZIO effect's return value. 

Another option I was thinking on was using streams and having a sink that has the unprocessed items as "leftovers".

### CODE REVIEW

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBQuery.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBQuery.scala:555]: could we put this in the companion object for `BatchWriteItem` and call it just `Response` like we do for `BatchGetItem.Response` ?

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBQuery.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBQuery.scala:564]: I could not find where this is set anywhere? I guess the code has not been called yet.

**Review by googley42 — APPROVED** [review]
Nice work! - I think this is a good approach. Just one minor comment.
For testing unprocessed items we may be able to use the built in service mocks shipped with `zio-aws-dynamodb`

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBQuery.scala
Comment (nhyne) [reply dynamodb/src/main/scala/zio/dynamodb/DynamoDBQuery.scala:564]: Yea right now I don't have any callers for it. I was going to add a helper to set it. Maybe something like `withRetryAttempts(Int)` -- (maybe a good use case for zio-prelude for positive only numbers?) and `withRetryTimeout(Duration)`?

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:221]: I think having no tail recursion is OK here as the function returns an effect which makes it stack safe

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:322]: I think having no tail recursion is OK here as the function returns an effect which makes it stack safe

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:60]: I think you can get rid of `.provideLayer(Clock.live)` by adding a constructor param that gets injected during ZLayer construction using `ZLayer.fromService`

```scala
private[dynamodb] final case class DynamoDBExecutorImpl private (clock: Clock.Service, dynamoDb: DynamoDb.Services)
    extends DynamoDBExecutor {
```

and gets constructed

```scala
object DynamoDBExecutor {
  val live: URLayer[DynamoDb with Clock, Has[DynamoDBExecutor]] =
    ZLayer
      .fromServices[Clock.Service, DynamoDb.Service, DynamoDBExecutor]((clock, dynamo) =>
        DynamoDBExecutorImpl(clock, dynamo)
      )
```

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:122]: just curious why the `DynamoDBExecutorImpl` prefix was introduced?

File: dynamodb/src/test/scala/zio/dynamodb/LiveExecutorSpec.scala
Comment (googley42) [inline dynamodb/src/test/scala/zio/dynamodb/LiveExecutorSpec.scala:69]: nice!

**Review by googley42 — COMMENTED** [review]
looks good, some comments on how layer dependencies can be pushed all the way out

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (nhyne) [reply dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:122]: I added it to hold some of the conversion functions from our model to the zio-aws model. I am using them in the tests in this PR but am wondering if that's a good idea now or if I should just create the zio-aws object and pass it instead.

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (nhyne) [reply dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:61]: I updated this but am not sure about passing the clock to the effects that need it. I currently have a `<effect>.provideLayer(ZLayer.succeed(clock))`. Does that make sense?

File: dynamodb/src/it/scala/zio/dynamodb/LiveSpec.scala
Comment (googley42) [inline dynamodb/src/it/scala/zio/dynamodb/LiveSpec.scala:39]:
```suggestion
  private val clockLayer = ZLayer.identity[Has[Clock.Service]]
```
This will allow you to push the clock dependency all the way up to be then satisfied with `.provideCustomLayer`

File: dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala
Comment (googley42) [inline dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala:321]:
```suggestion
  private val clockLayer = ZLayer.identity[Has[Clock.Service]]
```

File: dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala
Comment (googley42) [inline dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala:325]:
```suggestion
    batchGetSuite.provideCustomLayer((mockedBatchGet ++ clockLayer) >>> DynamoDBExecutor.live),

```

File: dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala
Comment (googley42) [inline dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala:326]:
```suggestion
    batchWriteSuite.provideCustomLayer((mockedBatchWrite ++ clockLayer) >>> DynamoDBExecutor.live)
```

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:66]:
```suggestion
        doBatchGetItem(batchGetItem).mapError(_.toThrowable).provide(Has(clock))
```

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:68]:
```suggestion
        doBatchWriteItem(batchWriteItem).mapError(_.toThrowable).provide(Has(clock))
```

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [reply dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:124]: np, I was just asking if `DynamoDBExecutorImpl` was not imported for clarity - which I guess is the case so that's cool

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [reply dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:61]: changes look good, I have some more suggestions below to leverage `ZEnv`

Comment (nhyne) [reply dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:62]: Updated this to use `Has`

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (nhyne) [reply dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala]: I imported the object, not sure why I didn't do this the first time.

**Comment by nhyne** [comment]
I'm still going to wait on a review with John to see if he has other suggestions before marking this PR as ready to merge/review; my guess is he will/

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:345]: Nice. same comment as above re `isInstanceOf`

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:383]: so my 50 cents on this is `GetItemRequest` is from `zio-aws` and the `projectionExpression` member is a list of `String`. I feel this list of string representation is a separate concern to the `ProjectionExpression` trait in the DSL - what do you think?

File: dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala
Comment (googley42) [inline dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala:35]: ah, I did not spot this before. The existing `ExecutorSpec` is testing code in the DSL layer and providing a fake for the executor. The new `batchRetries` is testing code in the live executor `DynamoDBExecutorImpl` so should probably live in its own spec file as they are testing different modules. Probably the original `ExecutorSpec` should be re-named to something else like `BatchingDslSpec` (I am rubbish when it comes to naming stuff)

File: dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala
Comment (googley42) [inline dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala:298]: do we have a test for "retrys" > "MAX retrys in policy " ? I think is will test the `catchSome` logic on the negative path.

**Review by googley42 — APPROVED** [review]
a great improvement on the retry functionality. Some minor comments and suggestions

File: dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala
Comment (nhyne) [reply dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala:298]: We don't have a test for this yet but I'll add some

File: dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala
Comment (nhyne) [reply dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala:35]: Okay, I can move the original tests to a new file and keep these new tests in here

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala
Comment (nhyne) [reply dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:383]: I think it should stay outside of the `ProjectionExpression` trait. This is an old comment. I think I'll add a helper somewhere though since this same `mkString(", ")` is called in a few places and it will make future refactors easier.

File: dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala
Comment (nhyne) [reply dynamodb/src/test/scala/zio/dynamodb/ExecutorSpec.scala]: Actually I did have one for just the writing, I've added one success and one failure for get/write to make sure both `catchSome`'s are tested.

**Review by googley42 — APPROVED** [review]
Nice work! LGTM

**Pull request merged**
